### PR TITLE
[ctx_prof] test tool: generate ctxprof bistream from json

### PR DESCRIPTION
**Pull request by mtrofin**
This is a tool to simplify testing. It generates a valid contextual profile file from a json representation.

The tool is authored to allow for future evolution, e.g. if we want to support profile merging or other tasks, not necessarily scoped to testing.

Issue #89287

### CODE REVIEW

File: llvm/tools/llvm-ctxprof-util/llvm-ctxprof-util.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    if (auto E = convertFromJSON()) {
+      handleAllErrors(std::move(E), [&](const ErrorInfoBase &E) {
+        E.log(errs());
+        errs() << "\n";
+      });
+      return 1;
+    }
+    return 0;
+  }
+  return 1;
```
Comment (teresajohnson) [inline llvm/tools/llvm-ctxprof-util/llvm-ctxprof-util.cpp:147]: Probably want to emit an error if it isn't FromJSON

File: llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
+[
+  {
+    "Guid": 1000,
+    "Counters": [1, 2, 3],
+    "Callsites": [
+      [],
+      [
+        {
+          "Guid": 2000,
+          "Counters": [4, 5]
```
Comment (teresajohnson) [inline llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json:10]: I don't see these counter values in the tested output

File: llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      {
+        "Guid": 3000,
+        "Counters": [40, 50]
+      }
+    ]
+  ]
+  },
+  {
+    "Guid": 18446744073709551612,
+    "Counters": [5, 9, 10]
```
Comment (teresajohnson) [inline llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json:27]: I don't see these counter values in the tested output

File: llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      [
+        {
+          "Guid": 2000,
+          "Counters": [4, 5]
+        },
+        {
+          "Guid": 18446744073709551613,
+          "Counters": [6, 7, 8]
+        }
+    ],
```
Comment (teresajohnson) [inline llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json:16]: Nit, can you align the matching brackets?

File: llvm/tools/llvm-ctxprof-util/llvm-ctxprof-util.cpp (hunk 1 above)
Comment (mtrofin) [reply llvm/tools/llvm-ctxprof-util/llvm-ctxprof-util.cpp:147]: the flag mechanism takes care of that, added a test though.

File: llvm/test/tools/llvm-ctxprof-util/llvm-ctxprof-util.test [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+; RUN: llvm-ctxprof-util fromJSON --input %S/Inputs/valid.json -output %t/valid.bitstream
+; RUN: llvm-bcanalyzer --dump %t/valid.bitstream | FileCheck %s --check-prefix=VALID
+
+; EMPTY: <BLOCKINFO_BLOCK/>
+; EMPTY-NEXT: <Metadata NumWords=1 BlockCodeSize=2>
+; EMPTY-NEXT:   <Version op0=1/>
+; EMPTY-NEXT: </Metadata>
+
+; Note that uin64_t are printed as signed values by llvm-bcanalyzer
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/test/tools/llvm-ctxprof-util/llvm-ctxprof-util.test:13]: done

File: llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json (hunk 2 above)
Comment (mtrofin) [reply llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json:10]: I only tested the values for a context, switching to testing the whole output.

File: llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json (hunk 3 above)
Comment (mtrofin) [reply llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json:27]: fixed.

File: llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json (hunk 4 above)
Comment (mtrofin) [reply llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json:16]: fixed - the reference file is indented.

**Review by teresajohnson — APPROVED** [review]
lgtm with a few suggestions

File: llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json (hunk 4 above)
Comment (mtrofin) [reply llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json:16]: oh you meant the json not the xml output. sorry. Misread.

Comment (mtrofin) [reply llvm/test/tools/llvm-ctxprof-util/Inputs/valid.json:16]: used a formatter this time.

File: llvm/test/tools/llvm-ctxprof-util/llvm-ctxprof-util.test [hunk 6]
```diff
@@ -3,37 +3,17 @@
...
-; RUN: llvm-bcanalyzer --dump %t/valid.bitstream | FileCheck %s --check-prefix=VALID
+
+; For the valid case, check against a reference output.
+; Note that uint64_t are printed as signed values by llvm-bcanalyzer:
+;  * 18446744073709551613 in json is -3 in the output
+;  * 18446744073709551612 in json is -4 in the output
+; Also we have no callee/context at index 0, 2 callsites for index 1, and one for
+; index 2.
+; RUN: llvm-bcanalyzer --dump %t/valid.bitstream | head -27 > %t.valid.in
+; RUN: diff %S/Inputs/valid.expected %t.valid.in
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/test/tools/llvm-ctxprof-util/llvm-ctxprof-util.test:14]: ack.

File: llvm/tools/llvm-ctxprof-util/llvm-ctxprof-util.cpp [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+#include "llvm/Support/JSON.h"
+#include "llvm/Support/LLVMDriver.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+
+static cl::SubCommand FromJSON("fromJSON", "Convert from json");
+
+static cl::opt<std::string> InputFilename(
```
Comment (nico) [inline llvm/tools/llvm-ctxprof-util/llvm-ctxprof-util.cpp:31]: As mentioned in https://github.com/llvm/llvm-project/pull/89884#issuecomment-2244348117 , cl:: is incompatible with GENERATE_DRIVER. Either remove GENERATE_DRIVER, or move this to Opt.

Comment (mtrofin) [reply llvm/tools/llvm-ctxprof-util/llvm-ctxprof-util.cpp:31]: Ack, I'll take care of of it. Thanks.

**Pull request merged**